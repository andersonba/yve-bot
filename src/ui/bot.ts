import { YveBot } from '../core/bot';
import { Rule, Answer, ChatOptions, ChatMessageSource } from '../types';
import { ChatUI } from './ui';


export class YveBotChat {
  private handlers: { [handler: string]: Array<() => any> };

  public bot: YveBot;
  public options: ChatOptions;
  public UI: ChatUI;

  constructor(rules: Rule[], opts?: ChatOptions) {
    const DEFAULT_OPTS: ChatOptions = {
      target: 'body',
      inputPlaceholder: 'Type your message',
      inputPlaceholderSingleChoice: 'Choose an option above',
      inputPlaceholderMutipleChoice: 'Choose the options above',
      doneMultipleChoiceLabel: 'Done',
      andSeparatorText: 'and',
      submitLabel: 'Send',
      timestampable: false,
      timestampFormatter: ts => new Date(ts).toUTCString().slice(-12, -4),
      autoFocus: true,
    };

    this.handlers = {};
    this.options = Object.assign({}, DEFAULT_OPTS, opts);

    this.bot = new YveBot(rules, this.options.yveBotOptions);
    this.UI = new ChatUI(this.options);

    this.bot
      .on('start', (...args) => {
        document
          .querySelector(this.options.target)
          .appendChild(this.UI.chat);

        if (this.options.autoFocus) {
          const $input = this.UI.input;
          $input.focus();
        }

        this.dispatch('start', ...args);
      })
      .on('talk', (msg: string, rule: Rule) => {
        this.newMessage('BOT', msg, rule);
      })
      .on('typing', () => this.typing())
      .on('typed', () => this.typed())
      .on('storeChanged', (...args) => this.dispatch('storeChanged', ...args))
      .on('end', (...args) => this.dispatch('end', ...args));

    this.UI.form.addEventListener('submit', ev => {
      ev.preventDefault();
      const input = this.UI.input;
      const msg = input.value.trim();

      if (msg) {
        this.dispatch('reply', msg);
        this.bot.hear(msg);
        this.newMessage('USER', msg);
        input.value = '';
      }
      if (this.options.autoFocus) {
        input.focus();
      }
    });
  }

  start() {
    this.bot.start();
    return this;
  }

  on(evt: string, fn: (...args: any[]) => any): this {
    if (evt in this.handlers) {
      this.handlers[evt].push(fn);
    } else {
      this.handlers[evt] = [fn];
    }
    return this;
  }

  dispatch(name: string, ...args) {
    if (name in this.handlers) {
      this.handlers[name].forEach(fn => fn(...args));
    }
    return this;
  }

  typing() {
    this.UI.typing.classList.add('is-typing');
    return this;
  }

  typed() {
    this.UI.typing.classList.remove('is-typing');
    return this;
  }

  newMessage(source: ChatMessageSource, message: Answer | Answer[], rule?: Rule) {
    const { UI } = this;
    const sender = source === 'BOT' ? this.options.name : null;
    const thread = UI.createThread(source, UI.createTextMessage(message, sender));

    if (source === 'BOT') {
      switch (rule.type) {
        case 'SingleChoice':
        thread.appendChild(UI.createSingleChoiceMessage(rule, (label, value) => {
          this.bot.hear(value);
          this.dispatch('reply', value);
          this.newMessage('USER', label);
        }));
        break;

        case 'MultipleChoice':
        thread.appendChild(UI.createMultipleChoiceMessage(rule, (label, value) => {
          this.bot.hear(value);
          this.dispatch('reply', value);
          this.newMessage('USER', label);
        }));
        break;
      }
    }
    UI.appendThread(this.UI.conversation, thread);
    return this;
  }
}
