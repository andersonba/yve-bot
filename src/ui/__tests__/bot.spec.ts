import { loadYaml, sleep, getChatElements } from '@test/utils';
import { YveBotChat } from '../bot';

const OPTS = {
  yveBotOptions: {
    enableWaitForSleep: false,
  },
};

beforeEach(() => {
  document.body.innerHTML = '';
});

test('event binding', async () => {
  const rules = loadYaml(`
  - name: value
    type: String
  `);
  const onStart = jest.fn();
  const onStoreChanged = jest.fn();
  const onEnd = jest.fn();
  const onEndCopy = jest.fn();
  const onReply = jest.fn();

  new YveBotChat(rules, OPTS)
    .on('start', onStart)
    .on('storeChanged', onStoreChanged)
    .on('end', onEnd)
    .on('end', onEndCopy)
    .on('reply', onReply)
    .start();
  const { input, submit } = getChatElements();

  const msg = 'ok';
  const output = {value: msg};
  const session = 'session';

  expect(onStart).toBeCalledWith(session);

  await sleep();

  input.value = msg;
  submit.click();

  await sleep();

  expect(onReply).toBeCalledWith(msg);
  expect(onStoreChanged).toBeCalledWith({ output, currentIdx: 1, waitingForAnswer: false }, session);
  expect(onEnd).toBeCalledWith(output, session);
  expect(onEndCopy).toBeCalledWith(output, session);
});

describe('DOM behaviors', () => {
  test('initial elements', async () => {
    const rules = loadYaml(`
    - message: Hello
    `);

    new YveBotChat(rules, OPTS).start();
    const { target, conversation, form, input, submit, getTyping } = getChatElements();

    await sleep();

    expect(conversation).not.toBeNull();
    expect(getTyping()).not.toBeNull();
    expect(form).not.toBeNull();
    expect(input).not.toBeNull();
    expect(submit).not.toBeNull();
    expect(target).toMatchSnapshot();
  });

  test('bot with name', async () => {
    const rules = loadYaml(`
    - message: Hello
    `);
    new YveBotChat(rules, Object.assign({ name: 'YveBot' }, OPTS)).start();
    const { target, getSenderName } = getChatElements();

    await sleep();

    expect(getSenderName()).not.toBeNull();
    expect(target).toMatchSnapshot();
  });

  test('bot with timestamp', async() => {
    const _now = Date.now;
    Date.now = jest.fn().mockReturnValue(0);

    const rules = loadYaml(`
    - message: Hello
    `);
    new YveBotChat(rules, Object.assign({ name: 'Yvebot', timestampable: true }, OPTS)).start();
    const { target, getTimestamp } = getChatElements();

    await sleep();

    expect(getTimestamp()).not.toBeNull();
    expect(target).toMatchSnapshot();

    Date.now = _now;
  });

  test('user reply', async () => {
    const rules = loadYaml(`
    - message: Your name
      type: String
    `);
    new YveBotChat(rules, OPTS).start();
    const { target, input, submit, getMessages, getUserMessages, getBotMessages } = getChatElements();

    await sleep();

    input.value = 'James';
    submit.click();

    await sleep();

    expect(getMessages()).toHaveLength(2);
    expect(getUserMessages()).toHaveLength(1);
    expect(getBotMessages()).toHaveLength(1);

    expect(target).toMatchSnapshot();
  });

  test('user reply without message', async () => {
    const rules = loadYaml(`
    - message: Your name
      type: String
    `);
    new YveBotChat(rules, OPTS).start();
    const { input, submit, getUserMessages } = getChatElements();

    input.value = '';
    submit.click();

    await sleep();

    expect(getUserMessages()).toHaveLength(0);
  });

  test('user reply with single choice', async () => {
    const rules = loadYaml(`
    - message: Make your choice
      type: SingleChoice
      options:
        - label: One
        - value: Two
        - label: Three
          value:
        - label:
          value: Four
        - Five
    `);
    new YveBotChat(rules, OPTS).start();
    const { target, input, submit, getBubbleButtons, getUserMessages } = getChatElements();

    await sleep();

    // showing the options
    const bubbles = getBubbleButtons();
    expect(input.hasAttribute('disabled')).toBeTruthy();
    expect(input.placeholder).toContain('Choose an option');
    expect(submit.hasAttribute('disabled')).toBeTruthy();
    expect(bubbles).toHaveLength(5);
    expect(getUserMessages()).toHaveLength(0);
    expect(target).toMatchSnapshot();

    // sending answer
    bubbles[0].click();
    await sleep();

    expect(input.hasAttribute('disabled')).toBeFalsy();
    expect(input.placeholder).toBe('Type your message');
    expect(submit.hasAttribute('disabled')).toBeFalsy();
    expect(getBubbleButtons()).toHaveLength(0);
    expect(getUserMessages()).toHaveLength(1);
    expect(target).toMatchSnapshot();
  });

  test('user reply with single choice and no options', async () => {
    const rules = loadYaml(`
    - message: Make your choice
      type: SingleChoice
    `);
    new YveBotChat(rules, OPTS).start();
    const { getBubbleButtons } = getChatElements();
    await sleep();
    expect(getBubbleButtons()).toHaveLength(0);
  });

  test('user reply with multiple choice', async () => {
    const rules = loadYaml(`
    - message: Make your choice
      type: MultipleChoice
      options:
        - label: One
        - value: Two
        - Three
    `);
    new YveBotChat(rules, OPTS).start();
    const { input, submit, getBubbleButtons, getBubbleDone, getUserMessages } = getChatElements();

    await sleep();

    const bubbles = getBubbleButtons();
    const done = getBubbleDone();

    // showing the options
    expect(done).not.toBeNull();
    expect(input.hasAttribute('disabled')).toBeTruthy();
    expect(input.placeholder).toContain('Choose the options');
    expect(submit.hasAttribute('disabled')).toBeTruthy();
    expect(bubbles).toHaveLength(3);
    expect(getUserMessages()).toHaveLength(0);

    // select one
    bubbles[0].click();
    expect(bubbles[0].classList).toContain('selected');
    expect(bubbles[1].classList).not.toContain('selected');
    expect(getUserMessages()).toHaveLength(0);

    // unselect
    bubbles[0].click();
    expect(bubbles[0].classList).not.toContain('selected');
    expect(bubbles[1].classList).not.toContain('selected');
    expect(getUserMessages()).toHaveLength(0);

    // select all
    bubbles[0].click();
    bubbles[1].click();
    expect(bubbles[0].classList).toContain('selected');
    expect(bubbles[1].classList).toContain('selected');
    expect(getUserMessages()).toHaveLength(0);

    // sending answer
    done.click();
    expect(input.hasAttribute('disabled')).toBeFalsy();
    expect(input.placeholder).toBe('Type your message');
    expect(submit.hasAttribute('disabled')).toBeFalsy();
    expect(getBubbleButtons()).toHaveLength(0);
    expect(getUserMessages()).toHaveLength(1);
  });

  test('user reply with multiple choice and no options', async () => {
    const rules = loadYaml(`
    - message: Make your choice
      type: MultipleChoice
    `);
    new YveBotChat(rules, OPTS).start();
    const { getBubbleButtons } = getChatElements();
    await sleep();
    expect(getBubbleButtons()).toHaveLength(0);
  });

  test('bot typing', async () => {
    const rules = loadYaml(`
    - message: Hello
      delay: 10
    `);
    new YveBotChat(rules).start();

    await sleep(5);

    // typing
    const { getTyping } = getChatElements();
    expect(getTyping().classList).toContain('is-typing');

    await sleep(20);

    // typed
    expect(getTyping().classList).not.toContain('is-typing');
  });

  test('bot sleeping', async () => {
    const rules = loadYaml(`
    - message: Hello
      delay: 1
    `);
    new YveBotChat(rules).start();

    // sleeping
    const { getTyping, getBotMessages } = getChatElements();
    expect(getTyping().classList).not.toContain('is-typing');
    expect(getBotMessages()).toHaveLength(0);

    // typed
    await sleep(5);
    expect(getBotMessages()).toHaveLength(1);
  });

  test('autofocus', async () => {
    const rules = loadYaml(`
    - type: Any
    - message: Make your choice
      type: SingleChoice
      options:
        - One
        - Two
    `);
    new YveBotChat(rules, OPTS).start();
    const { input, submit, getBubbleButtons } = getChatElements();

    await sleep();

    // sending message
    expect(document.activeElement).toEqual(input);
    input.value = 'A testing message';
    submit.click();
    expect(document.activeElement).toEqual(input);

    await sleep();

    // clicking on bubble
    const bubbles = getBubbleButtons();
    bubbles[0].click();
    expect(document.activeElement).toEqual(input);
  });

  test('disabled autofocus', async () => {
    const rules = loadYaml(`
    - type: Any
    - message: Make your choice
      type: SingleChoice
      options:
        - One
        - Two
    `);
    new YveBotChat(rules, Object.assign({ autoFocus: false }, OPTS)).start();
    const { input, submit, getBubbleButtons } = getChatElements();

    await sleep();

    // sending message
    expect(document.activeElement).not.toEqual(input);
    input.value = 'A testing message';
    submit.click();
    expect(document.activeElement).not.toEqual(input);

    await sleep();

    // clicking on bubble
    const bubbles = getBubbleButtons();
    bubbles[0].click();
    expect(document.activeElement).not.toEqual(input);
  });
});
