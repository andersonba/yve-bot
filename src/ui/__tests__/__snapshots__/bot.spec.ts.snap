// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DOM behaviors bot with name 1`] = `
<body>
  <div
    class="yvebot-chat"
  >
    <ul
      class="yvebot-conversation"
    >
      <li
        class="yvebot-thread yvebot-thread-bot"
      >
        <div
          class="yvebot-bubble"
        >
          <div
            class="yvebot-sender"
          >
            YveBot
          </div>
          <div
            class="yvebot-message"
          >
            Hello
          </div>
        </div>
      </li>
      <li
        class="yvebot-thread yvebot-thread-bot yvebot-thread-typing"
      >
        <div
          class="yvebot-typing"
        >
          <span
            class="yvebot-typing-dot"
          />
          <span
            class="yvebot-typing-dot"
          />
          <span
            class="yvebot-typing-dot"
          />
        </div>
      </li>
    </ul>
    <form
      class="yvebot-form"
    >
      <input
        autocomplete="off"
        class="yvebot-form-input"
        placeholder="Type your message"
        type="text"
      />
      <button
        class="yvebot-form-submit"
        type="submit"
      >
        Send
      </button>
    </form>
  </div>
</body>
`;

exports[`DOM behaviors bot with timestamp 1`] = `
<body>
  <div
    class="yvebot-chat"
  >
    <ul
      class="yvebot-conversation"
    >
      <li
        class="yvebot-thread yvebot-thread-bot"
      >
        <div
          class="yvebot-bubble"
        >
          <div
            class="yvebot-sender"
          >
            Yvebot
          </div>
          <div
            class="yvebot-message"
          >
            Hello
          </div>
          <div
            class="yvebot-timestamp"
          >
            00:00:00
          </div>
        </div>
      </li>
      <li
        class="yvebot-thread yvebot-thread-bot yvebot-thread-typing"
      >
        <div
          class="yvebot-typing"
        >
          <span
            class="yvebot-typing-dot"
          />
          <span
            class="yvebot-typing-dot"
          />
          <span
            class="yvebot-typing-dot"
          />
        </div>
      </li>
    </ul>
    <form
      class="yvebot-form"
    >
      <input
        autocomplete="off"
        class="yvebot-form-input"
        placeholder="Type your message"
        type="text"
      />
      <button
        class="yvebot-form-submit"
        type="submit"
      >
        Send
      </button>
    </form>
  </div>
</body>
`;

exports[`DOM behaviors initial elements 1`] = `
<body>
  <div
    class="yvebot-chat"
  >
    <ul
      class="yvebot-conversation"
    >
      <li
        class="yvebot-thread yvebot-thread-bot"
      >
        <div
          class="yvebot-bubble"
        >
          <div
            class="yvebot-message"
          >
            Hello
          </div>
        </div>
      </li>
      <li
        class="yvebot-thread yvebot-thread-bot yvebot-thread-typing"
      >
        <div
          class="yvebot-typing"
        >
          <span
            class="yvebot-typing-dot"
          />
          <span
            class="yvebot-typing-dot"
          />
          <span
            class="yvebot-typing-dot"
          />
        </div>
      </li>
    </ul>
    <form
      class="yvebot-form"
    >
      <input
        autocomplete="off"
        class="yvebot-form-input"
        placeholder="Type your message"
        type="text"
      />
      <button
        class="yvebot-form-submit"
        type="submit"
      >
        Send
      </button>
    </form>
  </div>
</body>
`;

exports[`DOM behaviors user reply 1`] = `
<body>
  <div
    class="yvebot-chat"
  >
    <ul
      class="yvebot-conversation"
    >
      <li
        class="yvebot-thread yvebot-thread-bot"
      >
        <div
          class="yvebot-bubble"
        >
          <div
            class="yvebot-message"
          >
            Your name
          </div>
        </div>
      </li>
      <li
        class="yvebot-thread yvebot-thread-user"
      >
        <div
          class="yvebot-bubble"
        >
          <div
            class="yvebot-message"
          >
            James
          </div>
        </div>
      </li>
      <li
        class="yvebot-thread yvebot-thread-bot yvebot-thread-typing"
      >
        <div
          class="yvebot-typing"
        >
          <span
            class="yvebot-typing-dot"
          />
          <span
            class="yvebot-typing-dot"
          />
          <span
            class="yvebot-typing-dot"
          />
        </div>
      </li>
    </ul>
    <form
      class="yvebot-form"
    >
      <input
        autocomplete="off"
        class="yvebot-form-input"
        placeholder="Type your message"
        type="text"
      />
      <button
        class="yvebot-form-submit"
        type="submit"
      >
        Send
      </button>
    </form>
  </div>
</body>
`;

exports[`DOM behaviors user reply with single choice 1`] = `
<body>
  <div
    class="yvebot-chat"
  >
    <ul
      class="yvebot-conversation"
    >
      <li
        class="yvebot-thread yvebot-thread-bot"
      >
        <div
          class="yvebot-bubble"
        >
          <div
            class="yvebot-message"
          >
            Make your choice
          </div>
        </div>
        <div
          class="yvebot-message-bubbles"
        >
          <button
            class="yvebot-message-bubbleBtn"
            data-label="One"
            data-value="One"
          >
            One
          </button>
          <button
            class="yvebot-message-bubbleBtn"
            data-label="Two"
            data-value="Two"
          >
            Two
          </button>
          <button
            class="yvebot-message-bubbleBtn"
            data-label="Three"
            data-value=""
          >
            Three
          </button>
          <button
            class="yvebot-message-bubbleBtn"
            data-label=""
            data-value="Four"
          />
          <button
            class="yvebot-message-bubbleBtn"
            data-label="Five"
            data-value="Five"
          >
            Five
          </button>
        </div>
      </li>
      <li
        class="yvebot-thread yvebot-thread-bot yvebot-thread-typing"
      >
        <div
          class="yvebot-typing"
        >
          <span
            class="yvebot-typing-dot"
          />
          <span
            class="yvebot-typing-dot"
          />
          <span
            class="yvebot-typing-dot"
          />
        </div>
      </li>
    </ul>
    <form
      class="yvebot-form"
    >
      <input
        autocomplete="off"
        class="yvebot-form-input"
        disabled=""
        placeholder="Choose an option above"
        type="text"
      />
      <button
        class="yvebot-form-submit"
        disabled=""
        type="submit"
      >
        Send
      </button>
    </form>
  </div>
</body>
`;

exports[`DOM behaviors user reply with single choice 2`] = `
<body>
  <div
    class="yvebot-chat"
  >
    <ul
      class="yvebot-conversation"
    >
      <li
        class="yvebot-thread yvebot-thread-bot"
      >
        <div
          class="yvebot-bubble"
        >
          <div
            class="yvebot-message"
          >
            Make your choice
          </div>
        </div>
      </li>
      <li
        class="yvebot-thread yvebot-thread-user"
      >
        <div
          class="yvebot-bubble"
        >
          <div
            class="yvebot-message"
          >
            One
          </div>
        </div>
      </li>
      <li
        class="yvebot-thread yvebot-thread-bot yvebot-thread-typing"
      >
        <div
          class="yvebot-typing"
        >
          <span
            class="yvebot-typing-dot"
          />
          <span
            class="yvebot-typing-dot"
          />
          <span
            class="yvebot-typing-dot"
          />
        </div>
      </li>
    </ul>
    <form
      class="yvebot-form"
    >
      <input
        autocomplete="off"
        class="yvebot-form-input"
        placeholder="Type your message"
        type="text"
      />
      <button
        class="yvebot-form-submit"
        type="submit"
      >
        Send
      </button>
    </form>
  </div>
</body>
`;
